src/network/networkd-dhcp4.c

```diff
@@ -588,9 +649,25 @@ static int dhcp_lease_acquired(sd_dhcp_client *client, Link *link) {
         }
 
         r = dhcp4_update_address(link, &address, &netmask, lifetime);
+        if (r < 0)
+                return log_link_warning_errno(link, r, "Could not update IP address: %m");
+
+        return 0;
+}
+
+static int dhcp_lease_ip_change(sd_dhcp_client *client, Link *link) {
+        int r;
+
+        link->dhcp_lease_old = TAKE_PTR(link->dhcp_lease);
+
+        r = dhcp_lease_acquired(client, link);
         if (r < 0) {
-                log_link_warning_errno(link, r, "Could not update IP address: %m");
-                link_enter_failed(link);
+                /* On failure, first drop the new lease */
+                (void) dhcp_lease_lost(link);
+
+                /* Then drop the old lease. */
+                link->dhcp_lease = TAKE_PTR(link->dhcp_lease_old);
+                (void) dhcp_lease_lost(link);
```

Comment:
This removes the call to `link_enter_failed()`. On purpose?

When dhpc_lease_lost() is called twice, warning "DHCP lease lost" will be emitted twice, and various state changes will also happen twice. I think this needs to be refactored to avoid this.

Reply:
What about the call to `link_enter_failed()`, is it removed on purpose?

Reply:
When this fails, then dhcp_lease_acquired() already failed, and thus `link_enter_failed()` will called.